### PR TITLE
[scheduler] Several scope, readability and test improvements

### PR DESCRIPTION
**Pull request by anandbonde**
This PR covers a series of commits for:
- Refactors to improve scope
- Refactors and renames to improve readability
- Simplification of test checks

### CODE REVIEW

File: src/rust/scheduler/scheduler.rs [hunk 1]
```diff
@@ -339,7 +339,7 @@ mod tests {
             None => anyhow::bail!("insert() failed"),
         };
         let task_id2: u64 = handle2.get_task_id();
-        crate::ensure_neq!(task_id2, task_id + 1);
+        crate::ensure_neq!(task_id2, task_id);
```
Comment (ppenna) [inline src/rust/scheduler/scheduler.rs:360]: Thanks for fixing that. Important question is: why the test was passing previously?

File: src/rust/scheduler/scheduler.rs [hunk 2]
```diff
@@ -130,19 +130,20 @@ impl Scheduler {
 
     /// Insert a new task into our scheduler returning a handle corresponding to it.
     pub fn insert<F: Task>(&self, future: F) -> Option<TaskHandle> {
+        let mut task_ids: RefMut<HashMap<u64, usize>> = self.task_ids.borrow_mut();
+        // If the address space for task ids is close to half full, it will become increasingly difficult to avoid
+        // collisions, so we cap the number of tasks.
+        if task_ids.len() > MAX_NUM_TASKS {
+            panic!("Too many concurrent tasks");
+        }
+
```
Comment (ppenna) [inline src/rust/scheduler/scheduler.rs]: What is the motivation for this change?

File: src/rust/scheduler/scheduler.rs [hunk 3]
```diff
@@ -164,16 +164,17 @@ impl Scheduler {
...
         let (page, subpage_ix): (&WakerPageRef, usize) = {
-            let (pages_ix, subpage_ix) = self.get_page_indexes(pin_slab_index);
+            let (pages_ix, subpage_ix) = self.get_page_and_offset_indexes(pin_slab_index);
             (&pages[pages_ix], subpage_ix)
         };
         page.initialize(subpage_ix);
         Some(TaskHandle::new(task_id, pin_slab_index, page.clone()))
     }
 
-    /// Computes the page and page offset of a given task based on its total offset.
```
Comment (ppenna) [inline src/rust/scheduler/scheduler.rs:171]: Please, do not remove existing comments.

File: src/rust/scheduler/scheduler.rs [hunk 4]
```diff
@@ -164,16 +164,17 @@ impl Scheduler {
...
-    fn get_page_indexes(&self, index: usize) -> (usize, usize) {
-        (index >> WAKER_BIT_LENGTH_SHIFT, index & (WAKER_BIT_LENGTH - 1))
+    fn get_page_and_offset_indexes(&self, pin_slab_index: usize) -> (usize, usize) {
+        let page_index: usize= pin_slab_index >> WAKER_BIT_LENGTH_SHIFT;
+        let offset_index: usize= pin_slab_index & (WAKER_BIT_LENGTH - 1);
+        (page_index, offset_index)
```
Comment (ppenna) [inline src/rust/scheduler/scheduler.rs]: There is a missing space between `usize` and `=`

File: src/rust/scheduler/scheduler.rs [hunk 5]
```diff
@@ -137,108 +137,106 @@ impl Scheduler {
...
                     match poll_result {
-                        Poll::Ready(()) => pages[page_ix].mark_completed(subpage_ix),
+                        Poll::Ready(()) => page_refs[page_index].mark_completed(page_offset),
                         Poll::Pending => (),
                     }
                 }
             }
             // There is some dropped task in this page, so iterate through it.
-            if dropped != 0 {
-                // Handle dropped tasks only.
```
Comment (ppenna) [inline src/rust/scheduler/scheduler.rs:219]: Please, do not remove existing comments.

File: src/rust/scheduler/scheduler.rs [hunk 6]
```diff
@@ -177,6 +175,15 @@ impl Scheduler {
...
 
+    // Add new page(s) to hold this future's status if the current page is filled. This may result in addition of
+    // multiple pages because of the gap between the pin slab index and the current page index.
+    fn add_new_pages_to_accommodate_index_gap(&self, pin_slab_index: usize) {
+        let mut page_ref: RefMut<Vec<WakerPageRef>> = self.page_refs.borrow_mut();
+        while pin_slab_index >= page_ref.len() << WAKER_BIT_LENGTH_SHIFT {
+            page_ref.push(WakerPageRef::default());
+        }
+    }
+
```
Comment (ppenna) [inline src/rust/scheduler/scheduler.rs]: I don't see any reason why this couldn't be inline as it was before.

File: src/rust/scheduler/scheduler.rs [hunk 7]
```diff
@@ -130,22 +130,33 @@ impl Scheduler {
...
-        let mut rng: RefMut<SmallRng> = self.rng.borrow_mut();
         let task_name: String = future.get_name();
-
-        // The pin slab index is the total index which can be reverse-computed in a page index and an offset within the
-        // page.
+        // The pin slab index can be reverse-computed in a page index and an offset within the page.
         let pin_slab_index: usize = self.tasks.borrow_mut().insert(Box::new(future))?;
 
-        // Generate a new id. If the id is currently in use, keep generating until we find an unused id.
+        self.add_new_pages_to_accommodate_index_gap(pin_slab_index);
```
Comment (ppenna) [inline src/rust/scheduler/scheduler.rs:139]: Why was this added here?

File: src/rust/scheduler/scheduler.rs [hunk 8]
```diff
@@ -130,22 +130,33 @@ impl Scheduler {
...
+        page_ref.initialize(page_offset);
+
+        let task_id = self.get_new_task_id(pin_slab_index);
+
+        trace!("insert(): name={:?}, id={:?}, pin_slab_index={:?}", task_name, task_id, pin_slab_index);
+        Some(TaskHandle::new(task_id, pin_slab_index, page_ref.clone()))
+    }
+
+    // Generate a new id. If the id is currently in use, keep generating until we find an unused id.
+    fn get_new_task_id(&self, pin_slab_index: usize) -> u64 {
```
Comment (ppenna) [inline src/rust/scheduler/scheduler.rs:156]: Before, this happened earlier. Why change?

**Review by ppenna — CHANGES_REQUESTED** [review]
I'd like some clarifications.

File: src/rust/scheduler/scheduler.rs (hunk 1 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:360]: I believe it was just a matter of time that it would fail... :)

File: src/rust/scheduler/scheduler.rs (hunk 2 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs]: This check is done regardless of what happens in the rest of the function, so I moved the check up to fail fast and not do any other work if this condition is met.

File: src/rust/scheduler/scheduler.rs (hunk 4 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs]: I fixed this in a subsequent commit. Do you think it's better to alter a prior commit and re-push?

File: src/rust/scheduler/scheduler.rs (hunk 5 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:219]: I was mislead by the comment to think that there would be a check whether the task is dropped or not. But, this code only iterates over the `dropped` bits, so there is no need of a check.

File: src/rust/scheduler/scheduler.rs [hunk 9]
```diff
@@ -84,157 +84,175 @@ pub struct Scheduler {
...
 
-        // Generate a new id. If the id is currently in use, keep generating until we find an unused id.
-        let mut task_ids: RefMut<HashMap<u64, usize>> = self.task_ids.borrow_mut();
-        // If the address space for task ids is close to half full, it will become increasingly difficult to avoid
-        // collisions, so we cap the number of tasks at 16,000.
-        if task_ids.len() > MAX_NUM_TASKS {
-            panic!("Too many concurrent tasks");
-        }
-        let task_id: u64 = 'get_id: {
+        self.add_new_pages_to_accommodate_index_gap(pin_slab_index);
```
Comment (iyzhang) [inline src/rust/scheduler/scheduler.rs:139]: maybe name this: allocate_new_pages_up_to_waker_index or something like that

File: src/rust/scheduler/scheduler.rs [hunk 10]
```diff
@@ -84,157 +84,175 @@ pub struct Scheduler {
...
-        let mut pages: RefMut<Vec<WakerPageRef>> = self.pages.borrow_mut();
+        let mut page_refs: RefMut<Vec<WakerPageRef>> = self.page_refs.borrow_mut();
         let mut tasks: RefMut<PinSlab<Box<dyn Task>>> = self.tasks.borrow_mut();
 
         // Iterate through pages.
-        for page_ix in 0..pages.len() {
-            let (notified, dropped): (u64, u64) = {
-                let page: &mut WakerPageRef = &mut pages[page_ix];
-                (page.take_notified(), page.take_dropped())
+        for page_index in 0..page_refs.len() {
```
Comment (iyzhang) [inline src/rust/scheduler/scheduler.rs:204]: It's probably fine to treat these as: page in pages or waker_page in waker_pages

File: src/rust/scheduler/scheduler.rs (hunk 3 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:171]: I was hoping to make it clear in the function name so that we don't need the comment. Is it still ambiguous?

File: src/rust/scheduler/scheduler.rs (hunk 1 above)
Comment (iyzhang) [reply src/rust/scheduler/scheduler.rs:360]: I also wanted to check that the scheduler is not allocating task IDs sequentially. I suppose even with a random allocation, this might fail occasionally. Any better ideas for how to test this? Anand was the one that suggested that I write a test case for this.  :)

File: src/rust/scheduler/scheduler.rs [hunk 11]
```diff
@@ -71,9 +71,9 @@ pub struct Scheduler {
     /// Maps between externally meaningful ids and the index of the task in the slab.
     task_ids: Rc<RefCell<HashMap<u64, usize>>>,
     /// Holds the waker bits for controlling task scheduling.
-    pages: Rc<RefCell<Vec<WakerPageRef>>>,
+    page_refs: Rc<RefCell<Vec<WakerPageRef>>>,
```
Comment (iyzhang) [inline src/rust/scheduler/scheduler.rs:74]: I think it's fine to call this waker_pages. It is a reference but the fact that it is a reference isn't important, they are actually being allocated here.

**Review by iyzhang — APPROVED** [review]
LGTM. Would appreciate it if you can add a test to ensure our task ids do not follow a pre-defined pattern.

File: src/rust/scheduler/scheduler.rs (hunk 8 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:156]: There is no requisite temporal ordering within these operations, so I moved the operations to the closest place where they are used.

File: src/rust/scheduler/scheduler.rs (hunk 6 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs]: I'm moving as many lower levels of abstractions as possible to their specific scope so that the semantics of the calling function stand out clearly. This is a lower level abstraction that is complex enough that it deserves a separate logical block (a function in this case).

File: src/rust/scheduler/scheduler.rs (hunk 11 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:74]: I called it `page_refs` because we have defined separate WakerPage and WakerPageRef structures in code explicitly. They are not implicit refs as part of the language, so wanted to have that distinction. Also, there is functionality in the WakerPageRef's non-trivial `default()` implementation (which we use to allocate the pages here).

So, I thought it might be better to call them what they are. Do let me know if you think otherwise.

File: src/rust/scheduler/scheduler.rs (hunk 1 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:360]: Oh, I was just asking for a test covering this code :).

If we stick to a good/decent random number generator (RNG) then we don't need to be worried about this at all because that's a test for the RNG, and not for this code.

**Comment by anandbonde** [comment]
> LGTM. Would appreciate it if you can add a test to ensure our task ids do not follow a pre-defined pattern.

Just testing for uniqueness should be good enough for this code because testing for a pattern falls under the umbrella of the RNG that we depend on.

File: src/rust/scheduler/scheduler.rs (hunk 10 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:204]: I called it `page_refs` because we have defined separate WakerPage and WakerPageRef structures in code explicitly. They are not implicit refs as part of the language, so wanted to have that distinction. Also, there is functionality in the WakerPageRef's non-trivial `default()` implementation (which we use to allocate the pages here).

So, I thought it might be better to call them what they are. Do let me know if you think otherwise.

File: src/rust/scheduler/scheduler.rs (hunk 9 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs]: Just updated it to match up to that meaning. Let me know if that looks ok.

File: src/rust/scheduler/scheduler.rs (hunk 1 above)
Comment (ppenna) [reply src/rust/scheduler/scheduler.rs:360]: > I believe it was just a matter of time that it would fail... :)

It will not fail if and only if the number generator is somehow is generating numbers in sequence. We can check that by changing the seed number.  Could you verify? I would say that there is something else going on here.

File: src/rust/scheduler/scheduler.rs (hunk 3 above)
Comment (ppenna) [reply src/rust/scheduler/scheduler.rs:171]: Well, I have already expressed my opinion on this many times... I am not arguing against your renaming, but I am not ok with removing comments. I think that this is a middle ground, right?

File: src/rust/scheduler/scheduler.rs (hunk 4 above)
Comment (ppenna) [reply src/rust/scheduler/scheduler.rs]: This is ok.

File: src/rust/scheduler/scheduler.rs (hunk 5 above)
Comment (ppenna) [reply src/rust/scheduler/scheduler.rs:219]: The comment does not say "check for dropped tasks", it says to "handle" them. If the comment is not good as expressive as it could be, let's update it, not remove.

File: src/rust/scheduler/scheduler.rs (hunk 1 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:360]: I didn't understand your point. This test will only fail (in the current state without my change) if the 2 task ids are in sequence. It has nothing to do with the seed. Which scenario are you thinking about?

File: src/rust/scheduler/scheduler.rs (hunk 5 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:219]: The wording in this comment (the word only) led me to think that there would be a check, so I went searching for it. If we are changing the existing comment becuase it is not helpful, then removing it altogether makes sense. Do let me know if you have any suggestions for what needs to be commented at this place - especially when the comment 2 lines above explains what the code below is going to do.

File: src/rust/scheduler/scheduler.rs (hunk 3 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:171]: We should consider such opinions case by case and cannot treat them as thumb rules. But, I will keep this comment because it does add context about the total offset.

File: src/rust/scheduler/scheduler.rs (hunk 5 above)
Comment (iyzhang) [reply src/rust/scheduler/scheduler.rs:219]: I think we do not need to discuss this because we want to get rid of this. In fact if there is some way to mark the code as deprecated, that would be great.

File: src/rust/scheduler/scheduler.rs (hunk 11 above)
Comment (iyzhang) [reply src/rust/scheduler/scheduler.rs:74]: sure can we call them waker_refs or waker_page_refs then? The waker part is more important than the page part.

File: src/rust/scheduler/scheduler.rs (hunk 3 above)
Comment (iyzhang) [reply src/rust/scheduler/scheduler.rs:171]: I think we should leave comments for every function because we may want to generate rust docs in the future. In fact, I would advocate that we eventually document every input and output to every function. Sometimes I find myself searching the code for what the function actually outputs and it's really annoying.

File: src/rust/scheduler/scheduler.rs (hunk 1 above)
Comment (iyzhang) [reply src/rust/scheduler/scheduler.rs:361]: I think Pedro is wondering if the random number generator ever generates two numbers in sequence for a given seed and if it does, we should just use a different seed. I think it's pretty unlikely but not impossible that the RNG produces to numbers in sequence, so we should probably not test it. I agree with this change.

**Review by iyzhang — APPROVED** [review]

File: src/rust/scheduler/scheduler.rs (hunk 11 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:74]: Yes, that makes sense. I will do the change.

File: src/rust/scheduler/scheduler.rs (hunk 3 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:171]: Ok, I reverted the comment. We can discuss offline about more on this, if needed.

File: src/rust/scheduler/scheduler.rs (hunk 5 above)
Comment (anandbonde) [reply src/rust/scheduler/scheduler.rs:219]: Ok, I reverted the comment. We can discuss offline about more on deprecation.

**Review by ppenna — APPROVED** [review]
